homeassistant/components/ecobee/config_flow.py

```diff
@@ -0,0 +1,120 @@
+            # Config entry already exists, only one allowed.
+            return self.async_abort(reason="one_instance_only")
+
+        errors = {}
+        stored_api_key = self.hass.data[DATA_ECOBEE_CONFIG].get(CONF_API_KEY)
+
+        if user_input is not None:
+            # Use the user-supplied API key to attempt to obtain a PIN from ecobee.
+            self._ecobee = Ecobee(config={ECOBEE_API_KEY: user_input[CONF_API_KEY]})
+
+            if await self.hass.async_add_executor_job(self._ecobee.request_pin):
+                # We have a PIN; move to the next step of the flow.
+                return await self.async_step_authorize()
+            errors["base"] = "pin_request_failed"
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_API_KEY, default=stored_api_key): str}
+            ),
+            errors=errors,
+        )
+
+    async def async_step_authorize(self, user_input=None):
+        """Present the user with the PIN so that the app can be authorized on ecobee.com."""
+        errors = {}
+
+        if user_input is not None:
+            # Attempt to obtain tokens from ecobee and finish the flow.
+            if await self.hass.async_add_executor_job(self._ecobee.request_tokens):
+                # Refresh token obtained; create the config entry.
+                config = {
+                    CONF_API_KEY: self._ecobee.api_key,
+                    CONF_REFRESH_TOKEN: self._ecobee.refresh_token,
+                }
+                return self.async_create_entry(title=DOMAIN, data=config)
+            errors["base"] = "token_request_failed"
+
+        return self.async_show_form(
+            step_id="authorize",
+            errors=errors,
+            description_placeholders={"pin": self._ecobee.pin},
+        )
+
+    async def async_step_import(self, import_data):
+        """
+        Import ecobee config from configuration.yaml.
+
+        Triggered by async_setup only if a config entry doesn't already exist.
+        If ecobee.conf exists, we will attempt to validate the credentials
+        and create an entry if valid. Otherwise, we will delegate to the user
+        step so that the user can continue the config flow.
+        """
+        try:
+            legacy_config = await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(ECOBEE_CONFIG_FILENAME)
+            )
+            config = {
+                ECOBEE_API_KEY: legacy_config[ECOBEE_API_KEY],
+                ECOBEE_REFRESH_TOKEN: legacy_config[ECOBEE_REFRESH_TOKEN],
+            }
+        except (HomeAssistantError, KeyError):
+            _LOGGER.debug(
+                "No valid ecobee.conf configuration found for import, delegating to user step"
+            )
+            return await self.async_step_user(
+                user_input={
+                    CONF_API_KEY: self.hass.data[DATA_ECOBEE_CONFIG].get(CONF_API_KEY)
```

Comment:
What happens if the api key is `None`? Should we check for that before passing the key to the user step?

Reply:
If the key is none (i.e. hasn't been supplied in configuration.yaml), the user step will attempt to validate it, fail, and the user step form will be shown to the user to start the flow fresh. For that reason I don't think it's necessary to validate that the api_key is not none in the import flow step.

Reply:
The same is true if the user supplies an invalid key in configuration.yaml - user step will attempt to validate and fail and the form will be shown. As I see it None is just a subspecies of invalid key that we are already equipped to handle.